mvc/src/main/java/cz/muni/fi/mvc/controllers/ComponentController.java

```diff
@@ -0,0 +1,177 @@
+        ComponentDTO foundComponent = componentFacade.findComponentByName(formBean.getName());
+
+        if (foundComponent != null && !foundComponent.getId().equals(formBean.getId())) {
+            redirectAttributes.addFlashAttribute("alert_warning", "Component with name '" + formBean.getName() + "' already exists");
+            return "redirect:" + uriBuilder.path("/components/edit/{id}").buildAndExpand(id).encode().toUriString();
+        }
+
+        if (bindingResult.hasErrors()) {
+            for (ObjectError ge : bindingResult.getGlobalErrors()) {
+                log.trace("ObjectError: {}", ge);
+            }
+            for (FieldError fe : bindingResult.getFieldErrors()) {
+                model.addAttribute(fe.getField() + "_error", true);
+                log.trace("FieldError: {}", fe);
+            }
+
+            model.addAttribute("componentEdit", formBean);
+            return "/components/edit";
+        }
+
+        log.debug("[COMPONENT] Update: {}", formBean);
+        model.addAttribute("componentTypeSelect", Arrays.asList(ComponentType.values()));
+        ComponentDTO result = componentFacade.updateComponent(formBean);
+
+        redirectAttributes.addFlashAttribute("alert_success", "Component '" + result.getName() + "' was updated");
+        return "redirect:" + uriBuilder.path("/components/read/{id}").buildAndExpand(id).encode().toUriString();
+    }
+
+    @RequestMapping(value = "/delete/{id}", method = RequestMethod.POST)
+    public String delete(@PathVariable long id, Model model, HttpServletRequest request, UriComponentsBuilder uriBuilder, RedirectAttributes redirectAttributes) {
+
+        String res = Tools.redirectNonTestDriver(request, uriBuilder, redirectAttributes);
+        if(res != null) return res;
+
+        ComponentDTO componentDTO = componentFacade.findComponentByID(id);
+        componentFacade.deleteComponent(componentDTO);
+        log.debug("delete component({})", id);
+        redirectAttributes.addFlashAttribute("alert_success", "Component '" + componentDTO.getName() + "' was deleted.");
+        return "redirect:" + uriBuilder.path("/components").build().toUriString();
+    }
+
+    @RequestMapping(value = "/read/{id}", method = RequestMethod.GET)
+    public String read(@PathVariable long id, Model model, UriComponentsBuilder uriBuilder, HttpServletRequest request) {
+        log.debug(" Read ({})", id);
+
+        model.addAttribute("component", componentFacade.findComponentByID(id));
+        return "components/read";
+    }
+
+    @RequestMapping(value="", method = RequestMethod.GET)
+    public String list(@Valid @ModelAttribute("filter")ComponentFilter filterBean, Model model, HttpServletRequest request, UriComponentsBuilder uriBuilder) {
+
+        ComponentFilter.ComponentFilterType filterNone = ComponentFilter.ComponentFilterType.NONE;
+
+        if (filterBean.getAvailable()) {
+            if (filterBean.getType() == filterNone || filterBean.getType() == null) {
```

Comment:
these blocks in if and else are almost the same. pls refactor without code repetition